magic_bridge/ic/src/magic_bridge/src/dab.rs

```diff
@@ -0,0 +1,136 @@
+use std::{str::{FromStr}, cell::RefCell, collections::HashMap};
+use ic_kit::{
+  candid::{ CandidType, Deserialize},
+  Principal,
+  ic,
+};
+
+use crate::types::*;
+use crate::factory::{CreateCanisterParam};
+
+const DAB_TOKEN_ADDRESS: &str = "zs2mv-bqaaa-aaaaa-aagla-cai";
+
+
+#[derive(CandidType, Deserialize, Clone, PartialEq, Debug)]
+pub enum DetailValue {
+    True,
+    False,
+    U64(u64),
+    I64(i64),
+    Float(f64),
+    Text(String),
+    Principal(Principal),
+    #[serde(with = "serde_bytes")]
+    Slice(Vec<u8>),
+    Vec(Vec<DetailValue>),
+}
+#[derive(CandidType, Deserialize)]
+pub struct DABParams {
+    pub name: String,
+    pub description: String,
+    pub thumbnail: String,
+    pub frontend: Option<String>,
+    pub principal_id: CanisterId,
+    pub details: Vec<(String, DetailValue)>,
+}
+
+#[derive(CandidType, Debug, Deserialize)]
+pub enum OperationError {
+    NotAuthorized,
+    NonExistentItem,
+    BadParameters,
+    Unknown(String),
+}
+
+pub type DABResponse = Result<(), OperationError>;
+
+#[derive(CandidType, Deserialize, Default, Clone)]
+pub struct DABHistory {
+    pub registered_canisters: RefCell<Vec<Principal>>,
```

Comment:
For magic bridge we use `thread_local!()`, we have a list of registered canisters, maybe we can use: https://github.com/Psychedelic/terabethia/blob/master/magic_bridge/ic/src/magic_bridge/src/magic.rs#L22 instead of `registered_canisters`. 

`failed_canisters` can also move under `MagicState`